src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -104,206 +69,46 @@ namespace OpenMS
-          }
-        }
-        else if (retention_time > peak_apex_rt_)
-        {
-          // end_time_at_5
-          if (intensity <= 0.05*peak_height_ &&
-            intensity_prev > 0.05*peak_height_)
-          {
-            delta_rt = retention_time - retention_time_prev;
-            delta_int = intensity_prev - intensity;
-            height_5 = 0.05*peak_height_ - intensity;
-            peakShapeMetrics.end_time_at_5 = retention_time - delta_int*delta_rt/height_5;
-          }
-          // start_time_at_10
-          if (intensity <= 0.1*peak_height_ &&
-            intensity_prev > 0.1*peak_height_)
-          {
-            delta_rt = retention_time - retention_time_prev;
-            delta_int = intensity_prev - intensity;
-            height_10 = 0.1*peak_height_ - intensity;
-            peakShapeMetrics.end_time_at_10 = retention_time - delta_int*delta_rt/height_10;
-          }
-          // end_time_at_50
-          if (intensity <= 0.5*peak_height_ &&
-          intensity_prev > 0.5*peak_height_)
-          {
-            delta_rt = retention_time - retention_time_prev;
-            delta_int = intensity_prev - intensity;
-            height_50 = 0.5*peak_height_ - intensity;
-            peakShapeMetrics.end_time_at_50 = retention_time - delta_int*delta_rt/height_50;
-          }
-        }
-
-        // points across the peak
-        peakShapeMetrics.points_across_baseline ++;
-        if (intensity >= 0.5*peak_height_)
-        {
-          peakShapeMetrics.points_across_half_height ++;
-        }
-      }
-    }
-
-    // peak widths
-    peakShapeMetrics.width_at_5 = peakShapeMetrics.end_time_at_5 - peakShapeMetrics.start_time_at_5;
-    peakShapeMetrics.width_at_10 = peakShapeMetrics.end_time_at_10 - peakShapeMetrics.start_time_at_10;
-    peakShapeMetrics.width_at_50 = peakShapeMetrics.end_time_at_50 - peakShapeMetrics.start_time_at_50;
-    peakShapeMetrics.total_width = right - left;
-    peakShapeMetrics.slope_of_baseline = end_intensity - start_intensity;
-    peakShapeMetrics.baseline_delta_2_height = peakShapeMetrics.slope_of_baseline / peak_height_;
+    calculatePeakShapeMetrics_(chromatogram, left, right, peakShapeMetrics);
+  }
 
-    // other
-    peakShapeMetrics.tailing_factor = peakShapeMetrics.width_at_5 / std::min(peak_apex_rt_ - peakShapeMetrics.start_time_at_5, peakShapeMetrics.end_time_at_5 - peak_apex_rt_);
-    peakShapeMetrics.asymmetry_factor = std::min(peak_apex_rt_ - peakShapeMetrics.start_time_at_10, peakShapeMetrics.end_time_at_10 - peak_apex_rt_) / std::max(peak_apex_rt_ - peakShapeMetrics.start_time_at_10, peakShapeMetrics.end_time_at_10 - peak_apex_rt_);
+  void PeakIntegrator::calculatePeakShapeMetrics(
+    const MSSpectrum& spectrum,
+    const double& left,
+    const double& right,
+    PeakShapeMetrics_& peakShapeMetrics
+  )
+  {
+    calculatePeakShapeMetrics_(spectrum, left, right, peakShapeMetrics);
```

Comment:
Doesn't the use of templates remove the need for this?  I.e., shouldn't either a MSSpectrum or MSChromatogram work?

Reply:
I'll try this out

Reply:
The generics were _protected_ so it wouldn't be possible to call them directly.

If I move them to _public_, then it works.

I guess it is safer to keep the template _protected_, and expose specialized _public_ methods so that if the user uses unexpected inputs types, the code would fail fast. Also, I'm not sure about what one needs to do with templates in pyOpenMS. :laughing: